### ISSUE TITLE
Replace capture syntax with `as` keyword

### ISSUE THREAD

**Issue by BanchouBoo** [ziglang/zig#9886]
The current syntax for captures is, in my opinion, untintuitive both to read and to write, especially for switch cases (I often find myself confused about where the capture is supposed to go there, before or after the `=>`), it'd be clearer to read and more comfortable to write if the captured value was defined right where you use the source value, hence `as`.

Some before and after examples:

If optionals:
```zig
const optional_number: ?usize = 12;
// current
if (optional_number) |number| {
    ...
}
// new
if (optional_number as number) {
    ...
}
```

While optional iterator:
```zig
var iter = iterator();
// current
while (iter.next) |value| {
    ...
}
// new
while (iter.next() as value) {
    ...
}
```

Switch cases:
```zig
const some_union: union(enum) {
    int: u32,
    float: f32,
} = .{ .int = 12 };
// current
switch (some_union) {
    .int => |i| ...,
    .float => |f| ...,
}
// new
switch (some_union) {
    .int as i => ...,
    .float as f => ...,
}
```

For loop:
```zig
const items = .{1, 2, 3, 4};
// current
for (items) |item| {
    ...
}
// new
for (items as item) {
    ...
}
```

If error union:
```zig
const error_union = someErrorableFunction();
// current
if (error_union) |value| {
    ....
} else |err| {
    ...
}
// new
if (error_union as value) {
    ....
} else (as err) {
    ...
}
```

Catch error union
```zig
// current
const value = someErrorableFunction() catch |err| ...;
// new
const value = someErrorableFunction() catch (as err) ...;
```

For #7257 I'm unsure if it would be better to do `thing thing thing capture capture capture` or `thing capture thing capture thing capture` so a few ways that could look are:
```zig
for (items_a, items_b, items_c as a, b, *c) {
   ...
}
for (items_a as a and items_b as b and items_c as *c) {
    ...
}
for (items_a as a, items_b as b, items_c as *c) {
    ...
}
```

It could also make the implementation of things like multi-capture If statements clearer, though that probably belongs in it's own proposal.

The only cons that comes to mind are that it becomes more difficult to visually scan through code to see where captures are and it may cause confusion between the `as` keyword and the `@as` builtin (perhaps there's a discussion to be had about using the `as` keyword as a replacement for `@as` as well, e.g. `const value = 12 as u32;` but that doesn't quite feel right to me and also doesn't seem like something other people would be on board with, but I'm not sure).

It's also quite possible that most people would rather the distinct separation and syntax that the current status quo provides and I'm just a weirdo shouting into the sky about things nobody else cares about, but I suppose we'll see.

**Comment by MasterQ32** [comment]
How would this look with this syntax:
```zig
for(list) |item, index| {

}
switch(tagged) {
    .a, .b => |int| { },
    .c, .d => |flt| { },
    else => |dat| { },
}
```

The capture syntax is nice as it allows having several captures at any place without looking odd, while the `as`-variant will definitly have weird precedence rules.

The catch is also amgigius:
```
const value = someErrorableFunction() catch |err| (5 - mapToInt(err));
// vs
const value = someErrorableFunction() catch (as err) (5 - mapToInt(err));
```
this immediate feels like it cannot be properly expressed with a PEG grammar and is hard to parse in general, imho it's also confusing why this is the only place in the language where two closing parens can follow each other, but it is *not* a function call

**Comment by Mouvedia** [comment]
I agree with @MasterQ32: @BanchouBoo how would you achieve #7257 using `as`?

**Comment by BanchouBoo** [comment]
@Mouvedia I already gave some suggestions in my initial post
